node/node.go

```diff
@@ -93,50 +65,73 @@ func run(ctx context.Context) (err error) {
 	mpool := mempool.New(mempooldb)
 
 	eventBus := bfttypes.NewEventBus()
-
-	logger := tmlog.NewTMLogger(tmlog.NewSyncWriter(os.Stdout))
-	ethAPI := struct {
-		*eth.ChainID
-		*eth.BlockByNumber
-		*eth.BlockByHash
-	}{
-		ChainID:       eth.NewChainID(g.ChainID.HexBig()),
-		BlockByNumber: eth.NewBlockByNumber(blockStore, rolluptypes.AdaptCosmosTxsToEthTxs),
-		BlockByHash:   eth.NewBlockByHash(blockStore, rolluptypes.AdaptCosmosTxsToEthTxs),
+	if err := eventBus.Start(); err != nil {
+		return fmt.Errorf("start event bus: %v", err)
 	}
-	n := newNodeService(
-		rpcee.NewEeRpcServer(ethHost, uint16(ethPort), []ethrpc.API{
-			{
-				Namespace: "eth",
-				Service:   ethAPI,
-			},
-		}, logger),
-		rpcee.NewEeRpcServer(engineHost, uint16(enginePort), []ethrpc.API{
-			{
-				Namespace: "engine",
-				Service: engine.NewEngineAPI(
-					builder.New(mpool, app, blockStore, txStore, eventBus, g.ChainID),
-					app,
-					rolluptypes.AdaptPayloadTxsToCosmosTxs,
-					blockStore,
-				),
-			},
-			{
-				Namespace: "eth",
-				Service:   ethAPI,
+	defer func() {
+		err = runAndWrapOnError(err, "stop event bus", eventBus.Stop)
+	}()
+
+	rpcServer := rpc.NewServer()
+	for _, api := range []rpc.API{
+		{
+			Namespace: "engine",
+			Service: engine.NewEngineAPI(
+				builder.New(mpool, app, blockStore, txStore, eventBus, n.genesis.ChainID),
+				app,
+				rolluptypes.AdaptPayloadTxsToCosmosTxs,
+				blockStore,
+			),
+		},
+		{
+			Namespace: "eth",
+			Service: struct {
+				*eth.ChainID
+				*eth.BlockByNumber
+				*eth.BlockByHash
+			}{
+				ChainID:       eth.NewChainID(n.genesis.ChainID.HexBig()),
+				BlockByNumber: eth.NewBlockByNumber(blockStore, rolluptypes.AdaptCosmosTxsToEthTxs),
+				BlockByHash:   eth.NewBlockByHash(blockStore, rolluptypes.AdaptCosmosTxsToEthTxs),
 			},
-		}, logger), eventBus)
+		},
+	} {
+		if err := rpcServer.RegisterName(api.Namespace, api.Service); err != nil {
+			return fmt.Errorf("register %s API: %v", api.Namespace, err)
+		}
+	}
 
-	if err := n.Start(); err != nil {
-		return fmt.Errorf("start node: %v", err)
+	httpServer := makeHTTPService(rpcServer, n.httpListener)
+	if err := httpServer.Start(); err != nil {
+		return fmt.Errorf("start http server: %v", err)
 	}
-	<-ctx.Done()
-	if err := n.Stop(); err != nil {
-		return fmt.Errorf("stop node: %v", err)
+	defer func() {
+		err = runAndWrapOnError(err, "stop http server", httpServer.Stop)
+	}()
+
+	wsServer := makeHTTPService(rpcServer.WebsocketHandler([]string{}), n.wsListener)
+	if err := wsServer.Start(); err != nil {
+		return fmt.Errorf("start websocket server: %v", err)
 	}
+	defer func() {
+		err = runAndWrapOnError(err, "stop websocket server", wsServer.Stop)
+	}()
+
+	<-ctx.Done()
 	return nil
 }
 
+func runAndWrapOnError(existingErr error, msg string, fn func() error) error {
+	if runErr := fn(); runErr != nil {
+		if existingErr == nil {
+			return runErr
+		}
+		runErr = fmt.Errorf("%s: %v", msg, runErr)
```

Comment:
I think this probably wants to be:

```suggestion
		runErr = fmt.Errorf("%s: %v", msg, runErr)
		if existingErr == nil {
			return runErr
		}
```

As is, if the pass function produces an error with no passed, bundled, error, the context string that describes what's being attempted gets lost.

Reply:
Great catch, thanks!